### PR TITLE
server: reload of grafana server certs when renewed without restart

### PR DESCRIPTION
**Pull request by chalapat**
**What is this feature?**
This feature is to support reloading the certificates of server when the certificates are renewed, instead of restarting the grafana.

**Why do we need this feature?**

with out this feature, grafana service needs to be restarted upon renewal of the certificates.  In some of the environments the certificates will be renewed daily. To take new certificates into effect, grafana needs to be restarted.

**Who is this feature for?**

For all who needs certificates to be reloaded up on renewal automatically.

**Special notes for your reviewer:**
Connection already established with old certificates needs to refresh the page upon reload. New connections will work with out grafana restart.

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Comment by chalapat** [comment]
Hi @diegommm @mildwonkey @yangkb09 ,
continous-integration/drone/pr stage is in pending state for long time. When i am trying to access, i am getting the following error "Build is blocked, please, contact repo admin in order to proceed".

Can you please help me out what is the issue.

Thanks

**Comment by DanCech** [comment]
Thanks for the submission @chalapat !

Being able to pick up a new cert on the fly sounds like a useful feature, but @diegommm and I are concerned about the overhead the current implementation would add to each request. That said, I think this could be implemented in a way that would avoid that overhead.

I suggest creating 2 properties on HTTPServer to hold the certificates and an RWMutex, then run a background goroutine that periodically checks whether the files have changed and updates the certificates (under Lock) if needed. We could potentially have a 3rd property to hold the mtime of the cert file so we can efficiently tell whether they are changed, or we could run readCertificates and compare the return cert list to the current one.

That would decouple checking for a new cert from answering a request, and allow us to only block requests long enough to update the property that holds the cert when we detect that the cert has changed.

It might also be good to add a config setting to control the check interval, and support setting it to 0 to disable the background cert check altogether for people who don't want/need it.

**Comment by chalapat** [comment]
Thanks @DanCech  @diegommm . I initially thought of adding such logic, then i have seen prometheus repository and saw there we no such checks put. 
i will add similar such logic

**Comment by chalapat** [comment]
@DanCech  @diegommm  I have two way of implementing;
using fsnotify:

- add certfile and keyfile to fsnotify watcher.
- when there is a change to the file an event from fsnotify is received, based on the event update the certs

using periodic watch:

- Based on the configured watchInterval,  check the modified times of the cert and key file.
- if modified time is not same as the  earlier one, read the cert and update the mtimes

Let me know which approach is better?

**Comment by diegommm** [comment]
Hi @chalapat! I believe the first approach is great and I personally like it, and we also considered it, but we leaned towards the second one since it avoids introducing a new direct dependency for an optional feature. I suggest using a periodic watch, but add a comment stating that we chose it to avoid importing fsnotify, but if we ever import fsnotify then we can switch the implementation to use it and have a more elegant solution.

Thank you for your contribution!

**Comment by chalapat** [comment]
@diegommm @DanCech  can you please enable the **continuous-integration/drone/pr** stage

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 1]
```diff
@@ -286,6 +286,17 @@ Path to the certificate file (if `protocol` is set to `https` or `h2`).
 
 Path to the certificate key file (if `protocol` is set to `https` or `h2`).
 
+### certs_watch_interval
+
+Periodicity at which `cert_key` and `cert_file` would be watched for changes.
+By default it is disabled. If enabled, then periodically `cert_key` and `cert_file`
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/_index.md:292]:
```suggestion
Disabled, by default. When enabled, `cert_key` and `cert_file`
```

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 2]
```diff
@@ -286,6 +286,17 @@ Path to the certificate file (if `protocol` is set to `https` or `h2`).
...
+are watched for changes, if there is change seen then those are certs are loaded automatically.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/_index.md:293]:
```suggestion
are periodically watched for changes. If there is a change, then those are certs are loaded automatically.
```

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 3]
```diff
@@ -286,6 +286,17 @@ Path to the certificate file (if `protocol` is set to `https` or `h2`).
...
+
+{{% admonition type="warning" %}}
+After the new certificates are loaded, connections which are established with old certificates
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/_index.md:296]:
```suggestion
After the new certificates are loaded, connections with old certificates
```

**Review by chri2547 — COMMENTED** [review]
@chalapat , please take a look at my wording suggestions to ensure I haven't changed the technical meaning you intend to communicate. Let me know if I'm off base and can work with you on finalizing. Thank you for the contribution!

**Comment by chalapat** [comment]
hi all, Something wrong happened to my push. it is showing too many files. i will close this and contribute via a different pull request. Sorry for the inconvenience.

**Pull request closed without merging**